src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -4632,35 +4675,41 @@ LogicalResult ONNXThresholdedReluOp::inferShapes(
   return emitError(NOT_IMPLEMENTED_MESSAGE);
 }
 
+//===----------------------------------------------------------------------===//
+// TopK
+//===----------------------------------------------------------------------===//
+
 LogicalResult ONNXTopKOp::verify() {
-  ONNXTopKOpAdaptor operandAdaptor = ONNXTopKOpAdaptor(*this);
-  // Get operands.
-  auto X = operandAdaptor.X();
-  auto K = operandAdaptor.K();
+  ONNXTopKOpAdaptor operandAdaptor(*this);
+  if (llvm::any_of(operandAdaptor.getOperands(),
+          [](const Value &op) { return !hasShapeAndRank(op); }))
+    return success(); // Won't be able to do any checking at this stage.
 
-  // Verify that axis value is in the valid range.
-  if (hasShapeAndRank(X)) {
-    ArrayRef<int64_t> shape = X.getType().cast<ShapedType>().getShape();
-    int64_t rank = shape.size();
-    int64_t axis = this->axis();
-    axis = axis < 0 ? axis + rank : axis;
-    if (axis < 0 || axis >= rank)
-      return emitOpError("axis must be in range [-rank, rank -1]");
-  }
+  Value X = operandAdaptor.X();
+  int64_t Xrank = X.getType().cast<ShapedType>().getRank();
+  int64_t axis = this->axis();
 
-  // Verify that K's rank must be zero or one.
-  if (hasShapeAndRank(K))
-    if (K.getType().cast<ShapedType>().getRank() > 1)
-      return emitOpError("K should have a rank of zero or one");
+  // axis attribute must be in the range [-r,r-1], where r = rank(X).
+  if (axis < -Xrank || axis >= Xrank)
+    return onnx_mlir::Diagnostic::emitAttributeOutOfRangeError(
+        *this->getOperation(), "axis", axis,
+        onnx_mlir::Diagnostic::Range<int64_t>(-Xrank, Xrank - 1));
+
+  // K's rank must be zero or one.
+  Value K = operandAdaptor.K();
+  int64_t KRank = K.getType().cast<ShapedType>().getRank();
+  if (KRank > 1)
+    return onnx_mlir::Diagnostic::emitOperandHasUnexpectedRankError(
+        *this->getOperation(), K, KRank, "< 2");
```

Comment:
K's rank does not depend on X's rank. Verifying this is doable without checking X's rank at the beginning of the function.

Reply:
moved lines 4759-4764 above the check for axis.